With over-parenthesized method call, E0615 suggests adding parentheses in the wrong place

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->

Given the following [code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=702c89fffada54a39f90d2ad99dae4bb): <!-- Please provide a link to play.rust-lang.org -->

```rust
fn main() {
    let a = Some(42);
    println!(
        "The value is {}.",
        (a.unwrap)()
    );
}
```

The current output is:

```
error[E0615]: attempted to take value of method `unwrap` on type `Option<{integer}>`
 --> src/main.rs:5:12
  |
5 |         (a.unwrap)()
  |            ^^^^^^ method, not a field
  |
help: use parentheses to call the method
  |
5 |         (a.unwrap)()()
  |                   ++
```

<!-- The following is not always necessary. -->
Ideally the output should look like:

```
help: remove these parentheses

(a.unwrap)()
^^^^^^^^^^ 
```

But at least adding parentheses right next to the method name would be an improvement:

```
help: use parentheses to call the method
  |
5 |         (a.unwrap())()
  |                  ++
```


<!--
If the problem is not self-explanatory, please provide a rationale for the
change.
-->

<!--
If dramatically different output is caused by small changes, consider also
adding them here.

If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions. The output might also be
different depending on the Edition.
-->

--- a/compiler/rustc_typeck/src/check/expr.rs
+++ b/compiler/rustc_typeck/src/check/expr.rs
@@ -1842,13 +1842,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             expr_t
         );
         err.span_label(field.span, "method, not a field");
-        let expr_is_call = if let hir::Node::Expr(parent_expr) =
-            self.tcx.hir().get(self.tcx.hir().get_parent_node(expr.hir_id))
-        {
-            matches!(parent_expr.kind, ExprKind::Call(..))
-        } else {
-            false
-        };
+        let expr_is_call =
+            if let hir::Node::Expr(hir::Expr { kind: ExprKind::Call(callee, _args), .. }) =
+                self.tcx.hir().get(self.tcx.hir().get_parent_node(expr.hir_id))
+            {
+                expr.hir_id == callee.hir_id
+            } else {
+                false
+            };
         let expr_snippet =
             self.tcx.sess.source_map().span_to_snippet(expr.span).unwrap_or(String::new());
         if expr_is_call && expr_snippet.starts_with("(") && expr_snippet.ends_with(")") {

--- a/compiler/rustc_typeck/src/check/expr.rs
+++ b/compiler/rustc_typeck/src/check/expr.rs
@@ -1842,7 +1842,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             expr_t
         );
         err.span_label(field.span, "method, not a field");
-        if !self.expr_in_place(expr.hir_id) {
+        let expr_is_call = if let hir::Node::Expr(parent_expr) =
+            self.tcx.hir().get(self.tcx.hir().get_parent_node(expr.hir_id))
+        {
+            matches!(parent_expr.kind, ExprKind::Call(..))
+        } else {
+            false
+        };
+        let expr_snippet =
+            self.tcx.sess.source_map().span_to_snippet(expr.span).unwrap_or(String::new());
+        if expr_is_call && expr_snippet.starts_with("(") && expr_snippet.ends_with(")") {
+            err.span_suggestion_short(
+                expr.span,
+                "remove wrapping parentheses to call the method",
+                expr_snippet[1..expr_snippet.len() - 1].to_owned(),
+                Applicability::MachineApplicable,
+            );
+        } else if !self.expr_in_place(expr.hir_id) {
             self.suggest_method_call(
                 &mut err,
                 "use parentheses to call the method",

--- a/compiler/rustc_typeck/src/check/expr.rs
+++ b/compiler/rustc_typeck/src/check/expr.rs
@@ -1853,10 +1853,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let expr_snippet =
             self.tcx.sess.source_map().span_to_snippet(expr.span).unwrap_or(String::new());
         if expr_is_call && expr_snippet.starts_with("(") && expr_snippet.ends_with(")") {
-            err.span_suggestion_short(
-                expr.span,
+            let after_open = expr.span.lo() + rustc_span::BytePos(1);
+            let before_close = expr.span.hi() - rustc_span::BytePos(1);
+            err.multipart_suggestion(
                 "remove wrapping parentheses to call the method",
-                expr_snippet[1..expr_snippet.len() - 1].to_owned(),
+                vec![
+                    (expr.span.with_hi(after_open), String::new()),
+                    (expr.span.with_lo(before_close), String::new()),
+                ],
                 Applicability::MachineApplicable,
             );
         } else if !self.expr_in_place(expr.hir_id) {

--- a/src/test/ui/typeck/issue-88803-call-expr-method.fixed
+++ b/src/test/ui/typeck/issue-88803-call-expr-method.fixed
@@ -4,6 +4,6 @@ fn main() {
     let a = Some(42);
     println!(
         "The value is {}.",
-        (a.unwrap)() //~ERROR [E0615]
+        a.unwrap() //~ERROR [E0615]
     );
 }

--- a/src/test/ui/typeck/issue-88803-call-expr-method.rs
+++ b/src/test/ui/typeck/issue-88803-call-expr-method.rs
@@ -1,0 +1,7 @@
+fn main() {
+    let a = Some(42);
+    println!(
+        "The value is {}.",
+        (a.unwrap)() //~ERROR [E0615]
+    );
+}

--- a/src/test/ui/typeck/issue-88803-call-expr-method.stderr
+++ b/src/test/ui/typeck/issue-88803-call-expr-method.stderr
@@ -1,0 +1,12 @@
+error[E0615]: attempted to take value of method `unwrap` on type `Option<{integer}>`
+  --> $DIR/issue-88803-call-expr-method.rs:5:12
+   |
+LL |         (a.unwrap)()
+   |         ---^^^^^^-
+   |         |  |
+   |         |  method, not a field
+   |         help: remove wrapping parentheses to call the method
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0615`.

--- a/src/test/ui/typeck/issue-88803-call-expr-method.stderr
+++ b/src/test/ui/typeck/issue-88803-call-expr-method.stderr
@@ -2,10 +2,13 @@ error[E0615]: attempted to take value of method `unwrap` on type `Option<{intege
   --> $DIR/issue-88803-call-expr-method.rs:7:12
    |
 LL |         (a.unwrap)()
-   |         ---^^^^^^-
-   |         |  |
-   |         |  method, not a field
-   |         help: remove wrapping parentheses to call the method
+   |            ^^^^^^ method, not a field
+   |
+help: remove wrapping parentheses to call the method
+   |
+LL -         (a.unwrap)()
+LL +         a.unwrap()
+   | 
 
 error: aborting due to previous error
 

--- a/src/test/ui/typeck/issue-88803-call-expr-method.stderr
+++ b/src/test/ui/typeck/issue-88803-call-expr-method.stderr
@@ -1,5 +1,5 @@
 error[E0615]: attempted to take value of method `unwrap` on type `Option<{integer}>`
-  --> $DIR/issue-88803-call-expr-method.rs:5:12
+  --> $DIR/issue-88803-call-expr-method.rs:7:12
    |
 LL |         (a.unwrap)()
    |         ---^^^^^^-

